includes/class-blocks.php

```diff
@@ -67,12 +67,50 @@ public static function register_blocks() {
 		);
 	}
 
+	/**
+	 * Get the user ID from a user string.
+	 *
+	 * @param string $user_string The user string. Can be a user ID, 'site', or 'inherit'.
+	 * @return int|null The user ID, or null if the 'inherit' string is not supported in this context.
+	 */
 	private static function get_user_id( $user_string ) {
 		if ( is_numeric( $user_string ) ) {
 			return absint( $user_string );
 		}
-		// any other non-numeric falls back to 0, including the `site` string used in the UI
-		return 0;
+
+		// if the user string is 'site', return the Blog User ID.
+		if ( 'site' === $user_string ) {
+			return User_Collection::BLOG_USER_ID;
+		}
+
+		// The only other value should be 'inherit', which means to use the query context to determine the User.
+		if ( 'inherit' !== $user_string ) {
+			return null;
+		}
+
+		// For a homepage/front page, if the Blog User is active, use it.
+		if ( ( is_front_page() || is_home() ) && ! is_user_type_disabled( 'blog' ) ) {
+			return 0;
+		}
+
+		// For other pages, the queried object will clue us in.
+		$queried_object = get_queried_object();
+		if ( ! $queried_object ) {
+			return null;
+		}
+
+		// If we're on a user archive page, use that user's ID.
+		if ( is_a( $queried_object, 'WP_User' ) ) {
+			return $queried_object->ID;
+		}
+
+		// For a single post, use the post author's ID.
+		if ( is_singular() && is_a( $queried_object, 'WP_Post' ) ) {
+			return $queried_object->post_author;
+		}
```

Comment:
This doesn't work in a query loop since the queried object will be the outer post. In my testing it was enough to just use `get_the_author()` which will have the right author inside the query loop and for the outer post.

```suggestion
		// For a single post, use the post author's ID.
		if ( is_a( $queried_object, 'WP_Post' ) ) {
			return get_the_author_meta( 'ID' );
		}
```

Reply:
Great find! Hadn't thought about loop stuff at all yet